### ISSUE TITLE
Quality issues?

### ISSUE THREAD

**Issue by bluenote10** [bmcfee/resampy#75]
Out of curiosity I did a test similar to [this](https://src.infinitewave.ca/) on the various resampling methods available in librosa. The test resamples an exponentially swept sine from 96kHz down to 44.1kHz. The test signal is 8 seconds long, and at around 7.2 seconds the swept sine passes the Nyquist frequency of the downsampled rate.

Considering that the resampy algorithm has a sound theoretical background straight from a JOS publication and is librosa's default I expected it to be very high quality. However what I got was somewhat surprising:

![resampy_best](https://user-images.githubusercontent.com/3620703/97913279-5a9faa80-1d4e-11eb-9b79-debe1e9b36bb.png)

![resampy_fast](https://user-images.githubusercontent.com/3620703/97913286-5d9a9b00-1d4e-11eb-95ff-a61302ef74d8.png)

![scipy signal resample](https://user-images.githubusercontent.com/3620703/97913301-612e2200-1d4e-11eb-9010-e1d6655fd6ca.png)

![scipy signal resample_poly](https://user-images.githubusercontent.com/3620703/97913315-668b6c80-1d4e-11eb-89dc-4a1c703c5ace.png)

![samplerate converters resample_sinc_best](https://user-images.githubusercontent.com/3620703/97913336-6be8b700-1d4e-11eb-8823-0ad6b8fce24a.png)

![samplerate converters resample_sinc_medium](https://user-images.githubusercontent.com/3620703/97913348-6f7c3e00-1d4e-11eb-9645-cedd894f6414.png)

![samplerate converters resample_sinc_fastest](https://user-images.githubusercontent.com/3620703/97913355-72772e80-1d4e-11eb-8789-1a2da82ac95d.png)

![samplerate converters resample_linear](https://user-images.githubusercontent.com/3620703/97913366-760ab580-1d4e-11eb-8fb1-4f612b6b5497.png)

![samplerate converters resample_zero_order_hold](https://user-images.githubusercontent.com/3620703/97913397-815de100-1d4e-11eb-9e43-76da390d9061.png)

<details>
<summary>full reproduction code</summary>

```python
import time

import numpy as np
import matplotlib.pyplot as plt

import librosa
import librosa.display


def exp_swept_sine(f1, f2, sr, amp=1.0, t=1.0):
    num_samples = int(t * sr)
    ts = np.arange(num_samples) / sr
    L = t / np.log(f2 / f1)
    wave = amp * np.sin(2.0 * np.pi * f1 * L * (np.exp(ts / L) - 1))
    return wave


def analyze_and_plot(wave, sr, method_name, runtime):
    hop_length = 256
    S = librosa.stft(wave, hop_length=hop_length)

    fig, ax = plt.subplots(1, 1, figsize=(20, 10))
    img = librosa.display.specshow(
        librosa.amplitude_to_db(np.abs(S), ref=np.max, amin=1e-10, top_db=180.0),
        y_axis='log',
        x_axis='time',
        sr=sr,
        ax=ax,
        hop_length=hop_length,
    )
    plt.colorbar(img, ax=ax)
    fig.suptitle("{} ({:.1f} ms)".format(method_name, runtime * 1000))
    fig.tight_layout()
    fig.savefig("/tmp/{}.png".format(method_name))
    plt.show()


def multi_check():
    sr1 = 96000
    sr2 = 44100
    wave1 = exp_swept_sine(f1=20, f2=sr1/2, sr=sr1, amp=0.5, t=8.0)

    methods = [(
        "resampy_best",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="kaiser_best")
    ), (
        "resampy_fast",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="kaiser_fast")
    ), (
        "scipy.signal.resample",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="scipy")
    ), (
        "scipy.signal.resample_poly",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="polyphase")
    ), (
        "samplerate.converters.resample_sinc_best",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="sinc_best")
    ), (
        "samplerate.converters.resample_sinc_medium",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="sinc_medium")
    ), (
        "samplerate.converters.resample_sinc_fastest",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="sinc_fastest")
    ), (
        "samplerate.converters.resample_linear",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="linear")
    ), (
        "samplerate.converters.resample_zero_order_hold",
        lambda: librosa.resample(wave1, sr1, sr2, res_type="zero_order_hold")
    )]

    for method_name, func in methods:
        t1 = time.time()
        wave2 = func()
        t2 = time.time()
        runtime = t2 - t1
        print("{:<50s} runtime: {:.1f} ms".format(method_name, runtime * 1000))
        analyze_and_plot(wave2, sr2, method_name, runtime)


if __name__ == "__main__":
    multi_check()
```

</details>

Any ideas why resampy shows such strong distortion?

**Comment by bmcfee** [comment]
Concerning indeed!  I remember doing some tests along these lines when initially developing resampy, and it was never quite as good as libsamplerate because I never bothered with the iterative optimization procedure for producing the window parameters.  In principle, we could do that at some point, and should get ~equivalent results.

The noise differences are in the -90dB range (from what I can tell), so not audible, but not great to see.  I wonder if there's some numerical precision factors at play here?  I haven't done a source-dive on libsamplerate, but it wouldn't surprise me if they're operating internally at higher precision than resampy is, and those round-off errors could add up.  (Just a guess!)

Failing that, it's entirely possible that we have some kind of a bug, but for now I'd chalk it up to less-than-optimal windowing.

**Comment by bmcfee** [comment]
After a bit of a dive into the libsamplerate filters, I think we can get some easy mileage out of changing the balance between the number of zero crossings (64 for kaiser_best) and the precision (number of interpolation points) in our filters.  For reference, the libsamplerate high-quality filter has a half-length of 340239, whereas our current "best" filter has about 1/10 as many: 32769 (64 zeros and precision=9).  Adding three bits of precision gets us in the same coefficient ballpark, but we can actually do better by going further with precision and removing some zeros from the filter.

Here's a prototype using 40 zeros, precision of 13 bits, beta=12.56ish, rolloff=0.90ish.  (Beta and rolloff optimized the method noted in #96 with some modifications to come.)  I've adjusted the colormap for these plots so that the midpoint at -120dB is black; anything tinted red is louder, blue is quieter.  The idea here is that artifacts in blue should be tolerable.
![image](https://user-images.githubusercontent.com/1190540/176211620-5936892f-9103-42d2-85e4-81d82079ffce.png)
compared to our current kaiser_best:
![image](https://user-images.githubusercontent.com/1190540/176211683-fe2165c7-cc5f-47e0-ab21-3700b656e449.png)
(note the similar runtimes)
and libsamplerate (much slower, but higher quality):
![image](https://user-images.githubusercontent.com/1190540/176211815-2ae8f373-a9d2-4987-b3c7-abd63e020ab2.png)
(Side note: resampy runtimes here are using the new parallel implementation, hence the speedup relative to other algorithms.)

TLDR is that we can bring the noise level down to the -120dB range without any appreciable loss in efficiency.  It remains to be seen how far we can push the low-quality filter, as the trade-off between length and interpolation might be qualitatively different in that regime.

**Comment by bmcfee** [comment]
After a bit of a dive into the libsamplerate filters, I think we can get some easy mileage out of changing the balance between the number of zero crossings (64 for kaiser_best) and the precision (number of interpolation points) in our filters.  For reference, the libsamplerate high-quality filter has a half-length of 340239, whereas our current "best" filter has about 1/10 as many: 32769 (64 zeros and precision=9).  Adding three bits of precision gets us in the same coefficient ballpark, but we can actually do better by going further with precision and removing some zeros from the filter.

Here's a prototype using 40 zeros, precision of 13 bits, beta=12.56ish, rolloff=0.90ish.  (Beta and rolloff optimized the method noted in #96 with some modifications to come.)  I've adjusted the colormap for these plots so that the midpoint at -120dB is black; anything tinted red is louder, blue is quieter.  The idea here is that artifacts in blue should be tolerable.
![image](https://user-images.githubusercontent.com/1190540/176211620-5936892f-9103-42d2-85e4-81d82079ffce.png)
compared to our current kaiser_best:
![image](https://user-images.githubusercontent.com/1190540/176211683-fe2165c7-cc5f-47e0-ab21-3700b656e449.png)
(note the similar runtimes)
and libsamplerate (much slower, but higher quality):
![image](https://user-images.githubusercontent.com/1190540/176211815-2ae8f373-a9d2-4987-b3c7-abd63e020ab2.png)
(Side note: resampy runtimes here are using the new parallel implementation, hence the speedup relative to other algorithms.)

TLDR is that we can bring the noise level down to the -120dB range without any appreciable loss in efficiency.  It remains to be seen how far we can push the low-quality filter, as the trade-off between length and interpolation might be qualitatively different in that regime.

**Comment by bmcfee** [comment]
Closing this one out now - 0.3 release improves things a bit, still not perfect, but within reasonable tolerances.
